### PR TITLE
Develop refactoring

### PR DESCRIPTION
**Pull request by oltionchampari**
Closes phovea/generator-phovea#425
### Developer Checklist (Definition of Done)

* [x] Descriptive title for this pull request provided (will be used for release notes later)
* [ ] All acceptance criteria from the issue are met
* [x] Branch is up-to-date with the branch to be merged with, i.e., develop
* [x] Code is cleaned up and formatted
* [ ] Code documentation written
* [x] Unit tests written
* [x] Build is successful
* [x] [Summary of changes](#summary-of-changes) written
* [ ] Wiki documentation written
* [x] Assign at least one reviewer
* [x] Assign at least one assignee
* [x] Add type label (e.g., *bug*, *enhancement*) to this pull request
* [x] Add next version label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)

### Summary of changes

- [x] utils:	
  - [x] split version to NpmUtils and PipUtils (make a VersionUtils for functions that can be used by both techniques)
  - [x] rename Repo to RepoUtils and make it static
  - [x] add pip to PipUtils
  - [ ] <del>add checkRequiredVersion(requriedNode, requiredNpm) to NpmUtils </del>  Add tests to test the check-node-version generator.

- [x] SpawnUtils	 
    - [x] abort(msg)
    - [x] spawn(cmd, argline, cwd)
    - [x] spawnOrAbort(cmd, argline,cwd)
    - [ ] <del>spawnCommand(command: string, args: string[], opt?: {}): any; --> check from generator</del> (It is not necessary since it is not used in any generator anymore. All generators use the sync version)
    - [x] spawnCommandSync(command: string, args: string[], opt?: {}): any; --> check from generator

- [ ] GeneratorUtils
    - [ ] checkVersion

- [x] WorkspaceUtils
     - [ ] <del> clonePlugins(log, plugins, ssh, reslolve, workspace, destinationPath) - from clone.</del> (This is part of the `clone generator` which is deprecated and does not work)
     - [x] resolveNeighbors(log, plugins, useSSH, types, shallow, destinationPath) - from clone
     - [x] cloneRepository(log, repoUrl, branch, extras, destinationPath, cwd) - from clone-repo (created a Util function called  `cloneRepo` that calls the `clone-repo generator` instead so we do not have to parse the arguments ourselves)
     - [x] add all of list-plugins

### CODE REVIEW

**Comment by anita-steiner** [comment]
@oltionchampari very, very good work :+1: 

could you also change all required with "xxx" to required with 'xxx'

File: base/BaseInitPluginGenerator.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+'use strict';
+const path = require('path');
+const fs = require('fs-extra');
+const chalk = require('chalk');
+const GeneratorUtils = require('../utils/GeneratorUtils');
+const BasePhoveaGenerator = require('../base/BasePhoveaGenerator');
+
+class BaseInitPluginGenerator extends BasePhoveaGenerator {
+
+    constructor(args, options, basetype) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply base/BaseInitPluginGenerator.js:10]: type refers to the different t

> @oltionchampari very, very good work 
> 
> could you also change all required with "xxx" to required with 'xxx'

Thank you for the feedback

File: generators/setup-workspace/index.js [hunk 2]
```diff
@@ -343,55 +267,32 @@ class Generator extends Base {
...
+  }
+
   writing() {
     this.hasErrors = false;
 
     return Promise.resolve(1)
-      .then(this._mkdir.bind(this, null))
+      .then(GeneratorUtils.mkdir(this.cwd))
       .then(this._getProduct.bind(this))
-      .then((product) => {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply generators/setup-workspace/index.js:352]: `this._getProduct.bind(this)` returns the parsed product - the repos. That step is included in the `this._getProduct.bind(this)`.

**Comment by oltionchampari** [comment]
@anita-steiner I have addressed your feedback, so you may take another look when you have time.

**Review by anita-steiner — COMMENTED** [review]
The following doesn't work as expected:
* init-setup: the workspace command is executed in the wrong directory; so the workspace is unusable at the moment
* app: throws an exception - what should this generator do?
* init-app-slib: throws an error: You don't seem to have a generator with the name “phovea:init-undefined” installed.
* why is the cwd a parameter for the generator and must be stored in the yo-rc file

**Comment by oltionchampari** [comment]
#### Documentation
* To check what genrator-phovea version is currently installed:
`npm ls -g --depth=0 generator-phovea`   ---> ` generator-phovea@6.0.1`

* To check wheather u are using the develop branch (linked)
`npm ls -g --depth=0 generator-phovea --link=true` --> `generator-phovea@6.0.2-SNAPSHOT`

We could integrate these commands as a `--version` flag in the generator.

**Comment by oltionchampari** [comment]
> The following doesn't work as expected:
> 
> * init-setup: the workspace command is executed in the wrong directory; so the workspace is unusable at the moment
> * app: throws an exception - what should this generator do?
> * init-app-slib: throws an error: You don't seem to have a generator with the name “phovea:init-undefined” installed.
> * why is the cwd a parameter for the generator and must be stored in the yo-rc file

* I removed the cwd from the yo-rc.json.
* The rest of the errors i  could not reproduce.

**Comment by anita-steiner** [comment]
> * To check what genrator-phovea version is currently installed:
>   `npm ls -g --depth=0 generator-phovea`   ---> ` generator-phovea@6.0.1`
> * To check wheather u are using the develop branch (linked)
>   `npm ls -g --depth=0 generator-phovea --link=true` --> `generator-phovea@6.0.2-SNAPSHOT`
> 
> We could integrate these commands as a `--version` flag in the generator.

I made a new issue for that: https://github.com/phovea/generator-phovea/issues/442

**Comment by anita-steiner** [comment]
I think there is again a problem with the folder structure: i executed init-app
image

The generator generates two folder: test and test-app

image

**Comment by anita-steiner** [comment]
in test-plugin there are only the following files:
- ISSUE_TEMPLATE
- LICENSE
- package.json
- README.md

**Pull request merged**